From 1e5dfdad7789497e7cd9636f0fe5b5e48045352d Mon Sep 17 00:00:00 2001
From: Willy Tarreau <w@1wt.eu>
Date: Sun, 7 Apr 2013 18:19:16 +0200
Subject: [PATCH 11/37] MINOR: session: only call http_send_name_header() when
 changing the server

Till now we used to call the function until the connection established, which
means that the header rewriting was performed for nothing upon each even (eg:
uploaded contents) until the server responded or timed out.

Now we only call the function when we assign the server.
---
 src/session.c | 19 ++++++++++---------
 1 file changed, 10 insertions(+), 9 deletions(-)

diff --git a/src/session.c b/src/session.c
index 8556284..d44583b 100644
--- a/src/session.c
+++ b/src/session.c
@@ -2194,21 +2194,22 @@ struct task *process_session(struct task *t)
 			 */
 			if (s->si[1].state != SI_ST_REQ)
 				sess_update_stream_int(s, &s->si[1]);
-			if (s->si[1].state == SI_ST_REQ)
+			if (s->si[1].state == SI_ST_REQ) {
 				sess_prepare_conn_req(s, &s->si[1]);
 
+				/* Now we can add the server name to a header (if requested) */
+				/* check for HTTP mode and proxy server_name_hdr_name != NULL */
+				if ((s->flags & SN_BE_ASSIGNED) &&
+				    (s->be->mode == PR_MODE_HTTP) &&
+				    (s->be->server_id_hdr_name != NULL)) {
+					http_send_name_header(&s->txn, s->be, objt_server(s->target)->id);
+				}
+			}
+
 			srv = objt_server(s->target);
 			if (s->si[1].state == SI_ST_ASS && srv && srv->rdr_len && (s->flags & SN_REDIRECTABLE))
 				http_perform_server_redirect(s, &s->si[1]);
 		} while (s->si[1].state == SI_ST_ASS);
-
-		/* Now we can add the server name to a header (if requested) */
-		/* check for HTTP mode and proxy server_name_hdr_name != NULL */
-		if ((s->flags & SN_BE_ASSIGNED) &&
-		    (s->be->mode == PR_MODE_HTTP) &&
-		    (s->be->server_id_hdr_name != NULL)) {
-			http_send_name_header(&s->txn, s->be, objt_server(s->target)->id);
-		}
 	}
 
 	/* Benchmarks have shown that it's optimal to do a full resync now */
-- 
1.8.1.5

